drivers/at24cxxx/at24cxxx.c

```diff
@@ -97,7 +97,31 @@ int _read(const at24cxxx_t *dev, uint32_t pos, void *data, size_t len)
         }
         xtimer_usleep(AT24CXXX_POLL_DELAY_US);
     }
+
     DEBUG("[at24cxxx] i2c_read_regs(): %d; polls: %d\n", check, polls);
+
+    return check;
+}
+
+static
+int _read(const at24cxxx_t *dev, uint32_t pos, void *data, size_t len)
+{
+    int check = 0;
+
+    while (len) {
+        size_t clen = MIN(len, DEV_PAGE_SIZE - MOD_POW2(pos, DEV_PAGE_SIZE));
```

Comment:
So this would also work?

```suggestion
        size_t clen = MIN(len, 256 - MOD_POW2(pos, 256));
```

Reply:
Yes, if it is written like this, with a limited read, it works:

```
int at24cxxx_read_page(const at24cxxx_t *dev, uint32_t page, uint32_t offset,
                       void *data, size_t len)
{
    size_t pos = page * DEV_PAGE_SIZE + offset;
    size_t clen = MIN(len, 256 - MOD_POW2(pos, 256));

    return at24cxxx_read(dev, pos, data, clen) == AT24CXXX_OK ? (int)clen : -EIO;
}
```

That is only a fix specific to this CPU's I2C implementation though, but it is an improvement nevertheless. We could only do it when e.g. `MODULE_SAM0_COMMON_PERIPH` is defined?

Reply:
Why do you need to subtract `MOD_POW2(pos, 256)`?

Also as @fabian18 pointed out the STM32 driver also has this limitation, so it might be more common. Reading up to 256 bytes at once is certainly an improvement to just reading 16 bytes (or failing).

Reply:
> Why do you need to subtract MOD_POW2(pos, 256)?

Oh I don't, that was just some bad copy pasta.

> Also as @fabian18 pointed out the STM32 driver also has this limitation, so it might be more common. Reading up to 256 bytes at once is certainly an improvement to just reading 16 bytes (or failing).

Other implementations may have even lower limits. Should we put a warning?

Reply:
> Other implementations may have even lower limits.

I doubt that.
I can see how some (HW) implementations would limit the transfer size to 1 byte length. I can't really see how less than 1 byte would make sense.

Reply:
> I doubt that.
I can see how some (HW) implementations would limit the transfer size to 1 byte length. I can't really see how less than 1 byte would make sense.

1 byte?

Reply:
But we are reading 256 bytes, not bits? I'm confused.

Reply:
If your `length` variable is 1 byte it can hold a maximum of 8 bits which can represent a value of up to 255

Reply:
Ah now I got it, you are talking about I2C implementations using e.g. `uint8_t` for the length, right?